### PR TITLE
feat(modules): add modules/search page

### PR DESCRIPTION
**Pull request by skaldarnar**
Uses [js-search](https://www.npmjs.com/package/js-search) to build a client-side search index and filter results.

Introduces a custom higher-order component to fetch location information from [@reach/router](https://www.npmjs.com/package/@reach/router) and parse it with [query-string](https://www.npmjs.com/package/query-string). The wrapper makes `location`, `navigate`, and the parsed `search` available to wrapped components.

In addition to the index search by search `term` the module search allows to filter by `tags` as well.

```
/modules/search/?tag=Library&tag=Assets
```

```
/modules/search/?term=rail
```

Based on 

- https://www.gatsbyjs.com/docs/adding-search-with-js-search
- https://medium.com/@chrisfitkin/how-to-get-query-string-parameter-values-in-gatsby-f714161104f
- https://www.dolthub.com/blog/2021-11-29-gatsby-search-and-pagination/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,6 +31,7 @@
     "gatsby-transformer-remark": "^6.3.2",
     "gatsby-transformer-sharp": "^5.3.1",
     "html-react-parser": "^1.2.7",
+    "js-search": "^2.0.0",
```
Comment (skaldarnar) [inline package.json:35]: The framework backing the search index. The searching happens all client-side, but I think we can statically compute the search index at build-time.

File: package.json [hunk 2]
```diff
@@ -9,6 +9,7 @@
   ],
   "author": "The Terasology Foundation",
   "dependencies": {
+    "@reach/router": "^1.3.4",
```
Comment (skaldarnar) [inline package.json:13]: Used for the higher-order component, but may not actually be needed. Need to double check...

File: package.json [hunk 3]
```diff
@@ -39,6 +41,7 @@
     "react-github-corner": "^2.5.0",
     "react-icons": "^3.7.0",
     "react-multi-carousel": "^2.6.3",
+    "react-select": "^5.7.0",
```
Comment (skaldarnar) [inline package.json:45]: Adds a _multi input_ for selection of tags from a pre-defined list.

File: src/components/SEO/SEO.jsx [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import React from "react";
 import urljoin from "url-join";
-import { useSiteMetadata } from "../../hooks/use-site-metadata";
+import { useSiteMetadata } from "../../hooks/useSiteMetadata";
```
Comment (skaldarnar) [inline src/components/SEO/SEO.jsx:3]: Not sure about our naming convention, but we mostly seem to use camel case for pages, components, etc.

File: src/components/common/withLocation.jsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import { Location } from "@reach/router";
+
+// Based on https://medium.com/@chrisfitkin/how-to-get-query-string-parameter-values-in-gatsby-f714161104f
+
+/* eslint-disable react/jsx-props-no-spreading - explicitly use property spreading for higher-order component (a component wrapping another component) */
+const withLocation = (ComponentToWrap) => (props) =>
```
Comment (skaldarnar) [inline src/components/common/withLocation.jsx]: I think this works for any component due to using the router directly, whereas with Gatsby's built-in feature only pages (and templates) can receive the `location` parameter.

File: src/components/modules/useJsSearch.jsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  // TODO: understand how JsSearch works and which parameters we want to use here
+  dataToSearch.indexStrategy = new JsSearch.AllSubstringsIndexStrategy();
+  dataToSearch.sanitizer = new JsSearch.LowerCaseSanitizer();
+
+  // Fields to search
+  dataToSearch.addIndex("name");
+  dataToSearch.addIndex(["moduleTxt", "description"]);
+  dataToSearch.addIndex(["moduleTxt", "tags"]);
+
+  dataToSearch.addDocuments(modules);
```
Comment (skaldarnar) [inline src/components/modules/useJsSearch.jsx:17]: This is building the search index on the fly, specifically tailored for modules. We could potentially do that build-time, although I don't know in which phase of Gatsby that would fit best (don't want to blow up the `createPages` phase again).

File: src/pages/modules/search.jsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import { Button, Col, Row, Form, FormGroup, Label, Input } from "reactstrap";
+import SEO from "../../components/SEO/SEO";
+import Layout from "../../layout";
+import ModuleListing from "../../components/modules/ModuleListing";
+import withLocation from "../../components/common/withLocation";
+import useJsSearch from "../../components/modules/useJsSearch";
+import ModuleSearchForm from "components/ModuleSearchForm/ModuleSearchForm";
+import Select from "react-select";
+
+// TODO: we really need to pass 'navigate' on or would the default from Gatsby suffice?
```
Comment (skaldarnar) [inline src/pages/modules/search.jsx]:
```suggestion
// TODO: move this into a separate component in `components/modules/`
```

File: src/pages/modules/search.jsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  const tags = searchParams.getAll("tag").filter((x) => x);
+  const term = searchParams.get("term") || "";
+
+  // filter based on searched term
+  const searchResults = term ? search(term) : allModules;
+  // filter based on tags
+  const modules = searchResults.filter((m) =>
+    tags.every((t) => m.moduleTxt.tags.includes(t))
+  );
```
Comment (skaldarnar) [inline src/pages/modules/search.jsx]: I'm not sure whether we should use some `useState`/`useEffect` here for the search params, etc.

File: src/pages/modules/search.jsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+  const allTags = allModules.reduce((acc, m) => {
+    m.moduleTxt.tags.forEach((t) => acc.add(t));
+    return acc;
+  }, new Set());
+
+  // build search index for all modules
+  const { search } = useJsSearch(allModules);
+
+  const [searchParams, setSearchParams] = useState(initialSearchParams);
+  // TODO: check whether we need this dirty flag (infinite rendering loops?)
```
Comment (skaldarnar) [inline src/pages/modules/search.jsx]: yes, we do...

**Comment by skaldarnar** [comment]
This adds the new pages to the navigation, but keeps the old files in place (but they are no longer linked, I think). We can remove the dead code either in this PR or in a follow up (module pages, old module pagination, hand-crafted module search index, `modules/` folder, possibly more).

File: src/components/common/withLocation.jsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
+import { Location } from "@reach/router";
+
+// Based on https://medium.com/@chrisfitkin/how-to-get-query-string-parameter-values-in-gatsby-f714161104f
+
+/* Explicitly use property spreading for higher-order component (a component wrapping another component). */
+/* eslint-disable react/jsx-props-no-spreading */
+function withLocation(ComponentToWrap) {
+  /* eslint-disable func-names */
+  return function (props) {
```
[the remark this answers is not in the document]
Comment (skaldarnar) [reply src/components/common/withLocation.jsx]: I'd argue that this rule does not add much value, but then I'm also used to lambda functions without names...

In this specific spot, I don't know what value giving this function a name would have (it's the function returned as a result from the higher-order function).

I've added a comment to disable the ESLint warning.

File: src/components/common/withLocation.jsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    return (
+      <Location>
+        {({ location, navigate }) => (
+          <ComponentToWrap
+            {...props}
```
[the remark this answers is not in the document]
Comment (skaldarnar) [reply src/components/common/withLocation.jsx:14]: This rule makes sense in general, but here we are explicitly creating a higher-order component for which we cannot know the wrapped component type, and therefore we cannot know the `props` that will be passed to it.

Therefore, I suggest to dismiss this alert, I've also added a comment to disable the ESLint check for this above.

**Review by jdrueckert — APPROVED** [review]

**Pull request merged**